pkg/network/tracer/kprobe/tracer.go

```diff
@@ -0,0 +1,478 @@
+//+build linux_bpf
+
+package kprobe
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode/runtime"
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	netebpf "github.com/DataDog/datadog-agent/pkg/network/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/network/ebpf/probes"
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/ebpf"
+	"github.com/DataDog/ebpf/manager"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	defaultClosedChannelSize = 500
+)
+
+type ConnectionTracer interface {
+	Start() (<-chan network.ConnectionStats, error)
+	Stop()
+	GetConnections(buffer []network.ConnectionStats, filter func(*network.ConnectionStats) bool) ([]network.ConnectionStats, error)
+	FlushPending() []network.ConnectionStats
+	Remove(conn *network.ConnectionStats) error
+	GetTelemetry() map[string]int64
+	GetMap(string) *ebpf.Map
+}
```

Comment:
A couple of thoughts/questions here:
* Shouldn't this `interface` be defined somewhere else (like `pkg/network/tracer`)?
* Given that the interface itself is quite complex, I'd document the interface methods as they serve as a guideline for alternative implementations.
* I worry that some of the implementations details are leaking into the interface. For example, is `FlushPending` really necessary or can we call it from `GetConnections` or something like that?

Reply:
`GetConnections` is for active connections, whereas `FlushPending` is for already closed connections that haven't been reported to userspace yet.

The problem is that the closed connections normally flow via the channel returned from `Start`. There is no way using that channel to ensure all the pending connections are processed before doing the delta state work.

Reply:
> Shouldn't this `interface` be defined somewhere else (like `pkg/network/tracer`)?

Moving to `pkg/network/tracer` would cause an import loop, since the `kprobe.NewTracer` function returns the interface. I could move the interface to its own package, and have the implementations one level deeper, similar to how `hash` is laid out in the stdlib.

Reply:
> I'd document the methods

Documentation added.

Reply:
> I could move the interface to its own package, and have the implementations one level deeper, similar to how hash is laid out in the stdlib.
I think that would probably make more sense since now the interface resides inside the kprobe package which just happens to be a possible implementation of it.